lib/spack/spack/concretize.py

```diff
@@ -362,7 +362,16 @@ def concretize_compiler(self, spec):
         # compiler_for_spec Should think whether this can be more
         # efficient
         def _proper_compiler_style(cspec, aspec):
-            return spack.compilers.compilers_for_spec(cspec, arch_spec=aspec)
+            compilers = spack.compilers.compilers_for_spec(
+                cspec, arch_spec=aspec
+            )
+            # If the spec passed as argument is concrete we want to check
+            # the versions match exactly
+            if (cspec.concrete and compilers and
+                cspec.version not in [c.version for c in compilers]):
+                return []
+
+            return compilers
 
         if spec.compiler and spec.compiler.concrete:
```

Comment:
> I would prefer if `zlib%gcc@6` could mean the same as  `zlib%gcc@6:6.999`…

I think that is achievable by modifying only this function, if desired. I think the issue is that a truncated version like `4.5` is considered concrete (this is a sensible assessment from the perspective of `Version`). This function could:

* Check for an exact match and return it if available
* Start with all `x.y.z` versions that are satisfied by the provided `x.y` version (e.g. use them to populate `matches` which is used below).

I don't think that's essential, but I do want to be clear that I think this is simpler than https://github.com/spack/spack/pull/14752#issuecomment-582449816 implies.

@alalazo do you have a preference of whether to try that out or stick with the current semantics?

Reply:
I would stick with the current semantics since it's the same used for packages:
```console
$ spack spec zlib@1
Input spec
--------------------------------
zlib@1

Concretized
--------------------------------
zlib@1%gcc@9.0.1+optimize+pic+shared arch=linux-ubuntu18.04-broadwell

$ spack spec zlib@1:
Input spec
--------------------------------
zlib@1:

Concretized
--------------------------------
zlib@1.2.11%gcc@9.0.1+optimize+pic+shared arch=linux-ubuntu18.04-broadwell

```

Reply:
IMO the semantics introduced here are different because in the case of packages it is "OK" to name a version that doesn't exist (in that case it means "go get the source for this remotely") but for compilers we are saying it's an error (as I say though that's OK with me).